common/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -416,180 +249,95 @@ __global__ void apply_kernel(
+            }
             __syncthreads();
 
-            logger.log_iteration(ibatch, iter, res_norms_shared_entry.values,
-                                 converged);
-
-            if (all_converged) {
+            // if (norms_res_sh[0] / norms_rhs_sh[0] < tol) {
+            if (stop.check_converged(norms_res_sh)) {
+                update_x_middle(nrows, alpha_sh[0], p_hat_sh, x_sh);
                 break;
             }
 
-
             // s_hat = precond * s
-            prec_shared.apply(gko::batch::to_const(s_shared_entry),
-                              s_hat_shared_entry);
-
+            prec_shared.apply(nrows, s_sh, s_hat_sh);
             __syncthreads();
 
-
             // t = A * s_hat
-            spmv_kernel(A_global_entry,
-                        gko::batch::to_const(s_hat_shared_entry),
-                        t_shared_entry);
+            single_spmv_kernel(a_global_entry, s_hat_sh, t_sh);
             __syncthreads();
 
             // omega = <t,s> / <t,t>
-            compute_omega(gko::batch::to_const(t_shared_entry),
-                          gko::batch::to_const(s_shared_entry),
-                          temp_shared_entry, omega_shared_entry, converged);
-
+            compute_omega(warp_grp, nrows, t_sh, s_sh, temp_sh[0],
+                          omega_sh[0] /*, converged*/);
             __syncthreads();
 
-
             // x = x + alpha*p_hat + omega *s_hat
             // r = s - omega * t
-            update_x_and_r(gko::batch::to_const(p_hat_shared_entry),
-                           gko::batch::to_const(s_hat_shared_entry),
-                           gko::batch::to_const(alpha_shared_entry),
-                           gko::batch::to_const(omega_shared_entry),
-                           gko::batch::to_const(s_shared_entry),
-                           gko::batch::to_const(t_shared_entry), x_shared_entry,
-                           r_shared_entry, converged);
-
-
+            update_x_and_r(nrows, p_hat_sh, s_hat_sh, alpha_sh[0], omega_sh[0],
+                           s_sh, t_sh, x_sh, r_sh /*, converged*/);
             __syncthreads();
 
-            compute_norm2<ValueType>(gko::batch::to_const(r_shared_entry),
-                                     res_norms_shared_entry,
-                                     converged);  // residual norms
-
-            __syncthreads();
-
-            // rho_old = rho_new
-            copy(gko::batch::to_const(rho_new_shared_entry),
-                 rho_old_shared_entry, converged);
+            if (threadIdx.x / config::warp_size == 0) {
+                single_compute_norm2(warp_grp, nrows, r_sh, norms_res_sh[0]);
+            }
+            //__syncthreads();
 
+            if (threadIdx.x == blockDim.x - 1) {
+                rho_old_sh[0] = rho_new_sh[0];
+            }
             __syncthreads();
         }
 
-        if (left_global_entry.values) {
-            batch_scale(right_global_entry, x_shared_entry);
```

Comment:
I guess now scaling is done via another kernel call ?

Reply:
Yes, it's now done in core using separate kernel calls. I think it's just more maintainable that way, and the overhead is not significant for this.